Properly resolve fullscreen API promises

nt1m opened the pull request:
#### 9696329f7af6ec156b6f45ae41d811da8fa0eeee
<pre>
Properly resolve fullscreen API promises
<a href="https://bugs.webkit.org/show_bug.cgi?id=248879">https://bugs.webkit.org/show_bug.cgi?id=248879</a>
rdar://103074395

Reviewed by Chris Dumez.

- exitFullscreen() promises should resolve in FullscreenManager::finishExitFullscreen().
- requestFullscreen() promises should resolve in FullscreenManager::willEnterFullscreen().

Store a pendingPromise field for async operations, and ensure there&apos;s only one promise at once, by rejecting the previous operation if it is still pending
(meaning the previous operation failed, since it is overriden by the new one).

We resolve this pending promise at the same time as emitting the fullscreen change/error events.

Removed mac-wk1 baselines that were due to this bug.

* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/selectors/modal-pseudo-class-expected.txt: Removed.
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/fullscreen/api/document-exit-fullscreen-nested-shadow-dom-expected.txt: Removed.
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/fullscreen/api/element-request-fullscreen-same-element-expected.txt: Removed.
* Source/WebCore/dom/DocumentFullscreen.cpp:
(WebCore::DocumentFullscreen::exitFullscreen):
(WebCore::DocumentFullscreen::webkitExitFullscreen):
* Source/WebCore/dom/DocumentFullscreen.h:
(WebCore::DocumentFullscreen::webkitExitFullscreen): Deleted.
* Source/WebCore/dom/FullscreenManager.cpp:
(WebCore::FullscreenManager::requestFullscreenForElement):
(WebCore::FullscreenManager::exitFullscreen):
(WebCore::FullscreenManager::willEnterFullscreen):
(WebCore::FullscreenManager::didExitFullscreen): Since 257456@main, cancelFullscreen now shares codepath with exitFullscreen, making the exitingDocument variable not needed.
(WebCore::FullscreenManager::notifyAboutFullscreenChangeOrError):
(WebCore::FullscreenManager::exitRemovedFullscreenElementIfNeeded):
(WebCore::FullscreenManager::clear):
(WebCore::FullscreenManager::dispatchFullscreenChangeEvents): Deleted.
* Source/WebCore/dom/FullscreenManager.h:
* Source/WebKit/WebProcess/InjectedBundle/API/gtk/DOM/WebKitDOMDocumentGtk.cpp:
(webkit_dom_document_webkit_exit_fullscreen):

Canonical link: <a href="https://commits.webkit.org/257668@main">https://commits.webkit.org/257668@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9ce25d868da5075a0d40dd01f2cbaaeddb63ffa9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/99643 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/8820 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/32724 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/109005 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/169238 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/103647 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/9406 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/86109 "Built successfully") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/92108 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/106914 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/105415 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/7090 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/90605 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/92108 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/88906 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/21962 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/92108 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/2644 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/23476 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/2588 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/45864 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5277 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/8726 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/42950 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/4451 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/dom/FullscreenManager.cpp:
```diff
@@ -237,10 +237,13 @@ void FullscreenManager::requestFullscreenForElement(Ref<Element>&& element, RefP
                 return;
             }
+            if (m_pendingPromise)
+                m_pendingPromise->reject(Exception { TypeError });
```
Please pass a second parameter to `Exception`. Should be a string that explains what went wrong.

cdumez commented on Source/WebCore/dom/FullscreenManager.cpp:
```diff
@@ -259,6 +262,10 @@ void FullscreenManager::cancelFullscreen()
         // by clearing the pending fullscreen element.
         m_pendingFullscreenElement = nullptr;
+        if (m_pendingPromise) {
+            m_pendingPromise->reject(Exception { TypeError });
```
Please pass a useful string as second parameter to Exception.

cdumez commented on Source/WebCore/dom/FullscreenManager.cpp:
```diff
@@ -343,15 +353,21 @@ void FullscreenManager::exitFullscreen()
         }
 
+        if (m_pendingPromise)
+            m_pendingPromise->reject(Exception { TypeError });
```
Please pass a useful string as second parameter.

cdumez commented on Source/WebCore/dom/FullscreenManager.cpp:
```diff
@@ -343,15 +353,21 @@ void FullscreenManager::exitFullscreen()
+        } else {
             finishExitFullscreen(document(), ExitMode::NoResize);
 
+            m_pendingPromise = WTFMove(promise);
```
Instead of duplicating the `m_pendingPromise = WTFMove(promise);` line, could we do it once before the `if (mode == ExitMode::Resize)` ?

cdumez commented on Source/WebCore/dom/FullscreenManager.cpp:
```diff
@@ -546,6 +557,15 @@ void FullscreenManager::dispatchFullscreenChangeEvents()
+            m_pendingPromise->reject(Exception { TypeError });
+        else if (!changeQueue.isEmpty())
+            m_pendingPromise->resolve();
+        m_pendingPromise = nullptr;
```
What about the case where both errorQueue and changeQueue are empty? Isn't it possible? If it isn't then it should be an assertion.

If it is possible, then I think we have an issue because you'll clear the pendingPromise pointer without resolving it, which would not be OK.

cdumez reviewed (approved):
LGTM with fixes.

nt1m committed:
Properly resolve fullscreen API promises
https://bugs.webkit.org/show_bug.cgi?id=248879
rdar://103074395

Reviewed by Chris Dumez.

- exitFullscreen() promises should resolve in FullscreenManager::finishExitFullscreen().
- requestFullscreen() promises should resolve in FullscreenManager::willEnterFullscreen().

Store a pendingPromise field for async operations, and ensure there's only one promise at once, by rejecting the previous operation if it is still pending
(meaning the previous operation failed, since it is overriden by the new one).

We resolve this pending promise at the same time as emitting the fullscreen change/error events.

Removed mac-wk1 baselines that were due to this bug.

* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/selectors/modal-pseudo-class-expected.txt: Removed.
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/fullscreen/api/document-exit-fullscreen-nested-shadow-dom-expected.txt: Removed.
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/fullscreen/api/element-request-fullscreen-same-element-expected.txt: Removed.
* Source/WebCore/dom/DocumentFullscreen.cpp:
(WebCore::DocumentFullscreen::exitFullscreen):
(WebCore::DocumentFullscreen::webkitExitFullscreen):
* Source/WebCore/dom/DocumentFullscreen.h:
(WebCore::DocumentFullscreen::webkitExitFullscreen): Deleted.
* Source/WebCore/dom/FullscreenManager.cpp:
(WebCore::FullscreenManager::requestFullscreenForElement):
(WebCore::FullscreenManager::exitFullscreen):
(WebCore::FullscreenManager::willEnterFullscreen):
(WebCore::FullscreenManager::didExitFullscreen): Since 257456@main, cancelFullscreen now shares codepath with exitFullscreen, making the exitingDocument variable not needed.
(WebCore::FullscreenManager::notifyAboutFullscreenChangeOrError):
(WebCore::FullscreenManager::exitRemovedFullscreenElementIfNeeded):
(WebCore::FullscreenManager::clear):
(WebCore::FullscreenManager::dispatchFullscreenChangeEvents): Deleted.
* Source/WebCore/dom/FullscreenManager.h:
* Source/WebKit/WebProcess/InjectedBundle/API/gtk/DOM/WebKitDOMDocumentGtk.cpp:
(webkit_dom_document_webkit_exit_fullscreen):

Canonical link: https://commits.webkit.org/257668@main
